File: units_java/src/test/java/com/units/tests/constants/TestWGS84.java
package com.units.tests.constants;

import static org.junit.Assert.assertTrue;

import org.junit.Test;
import static com.units.constants.Wgs84.*;

public class TestWGS84 {

	static final double error = 1e-10;
	
	private void assertAlmostEqual(double expected, double actual) {
		assertTrue(
				expected + " != " + actual,
			Math.abs(expected - actual) < error);
	}
	
	@Test
	public void testFlatenningToAxisesRelation() {
		double expected = flattening;
		double actual = 1 - semiMinorAxis.divide(semiMajorAxis).value();
		assertAlmostEqual(expected, actual);
	}
	
	@Test
	public void testFirstEccenticityToAxisesRelation() {
		double expected = firstEccentricity;
		double actual = Math.sqrt(1 - Math.pow(semiMinorAxis.divide(semiMajorAxis).value(), 2));
		assertAlmostEqual(expected, actual);
	}
	
	@Test
	public void testFirstEccenticityToFlatteningRelation() {
		double expected = firstEccentricity;
		double actual = Math.sqrt(flattening * (2 - flattening));
		assertAlmostEqual(expected, actual);
	}
	
	@Test
	public void testSecondEccenticityToAxisesRelation() {
		double expected = secondEccentricity;
		double actual = Math.sqrt(Math.pow(semiMajorAxis.divide(semiMinorAxis).value(), 2) - 1);
		assertAlmostEqual(expected, actual);
	}
	
	@Test
	public void testSecondEccenticityToFlatteningRelation() {
		double expected = firstEccentricity;
		double actual = secondEccentricity * (1 - flattening);
		assertAlmostEqual(expected, actual);
	}

}

File: units_cpp/src/unitsTest/cpp/constants/test_wgs84.cc
/*
 * test_wgs84.cc
 *
 *  Created on: Nov 1, 2017
 *      Author: sagis
 */

#include <gtest/gtest.h>
#include <units/constants/wgs84.h>

#include <cmath>

using namespace testing;
using namespace units::wgs84;

class TestWGS84 : public Test
{
protected:
	static constexpr double error = 1e-10;
};

TEST_F(TestWGS84, testFlatenningToAxisesRelation)
{
	ASSERT_NEAR(
		flattening,
		1 - semiMinorAxis.value() / semiMajorAxis.value(),
		error);
}

TEST_F(TestWGS84, testFirstEccenticityToAxisesRelation)
{
	ASSERT_NEAR(
		firstEccentricity,
		std::sqrt(1 - pow(semiMinorAxis.value() / semiMajorAxis.value(), 2)),
		error);
}

TEST_F(TestWGS84, testFirstEccenticityToFlatteningRelation)
{
	ASSERT_NEAR(
		firstEccentricity,
		std::sqrt(flattening * (2 - flattening)),
		error);
}

TEST_F(TestWGS84, testSecondEccenticityToAxisesRelation)
{
	ASSERT_NEAR(
		secondEccentricity,
		std::sqrt(pow(semiMajorAxis.value() / semiMinorAxis.value(), 2) - 1),
		error);
}

TEST_F(TestWGS84, testSecondEccenticityToFlatteningRelation)
{
	ASSERT_NEAR(
		secondEccentricity * (1 - flattening),
		firstEccentricity,
		error);
}
